concent_api/signing_service_integration_test_recovery.py

```diff
@@ -0,0 +1,246 @@
+            request_id=get_current_utc_timestamp(),
+        )
+
+        # Send message through wrapper and receive deserialized response.
+        response = send_message_to_middleman_and_receive_response(
+            message=golem_message_frame,
+            config=config,
+            concent_private_key=CONCENT_PRIVATE_KEY,
+            concent_public_key=CONCENT_PUBLIC_KEY,
+        )
+
+        # Check response.
+        assert_condition(type(response), GolemMessageFrame, f'Deserialized response type is {type(response)} instead of GolemMessageFrame.')
+        assert_condition(type(response.payload), SignedTransaction, f'Deserialized response payload type is {type(response.payload)} instead of SignedTransaction.')
+        assert_condition(response.request_id, golem_message_frame.request_id, f'Deserialized response request_id is {response.request_id} instead of {golem_message_frame.request_id}.')
+
+    except Exception as exception:
+        raise exception
+    finally:
+        middleman_process.kill()
+        signing_service_process.kill()
+
+
+@count_fails
+def test_case_1_middleman_recovery(config: ConfigParser, **kwargs: Any) -> None:
+    """
+    1. Spawn MiddleMan and SigningService processes.
+    2. Client sends GolemMessageFrame with correct TransactionSigningRequest to MiddleMan.
+    3. Middleman is restarted. The connection and latest message is lost.
+    4. Client sends GolemMessageFrame with correct TransactionSigningRequest to MiddleMan.
+    5. Client receives response for latest message.
+    """
+
+    try:
+        middleman_process = run_middleman()
+        signing_service_process = run_signing_service()
+
+        # Waiting for MiddleMan and SigningService to start.
+        sleep(10)
+
+        client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        client_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, True)
+
+        # Create GolemMessageFrame with correct TransactionSigningRequest.
+        golem_message_frame = create_golem_message_frame_with_correct_transaction_signing_request(
+            request_id=get_current_utc_timestamp(),
+        )
+
+        client_socket.connect(
+            (
+                config.get(Components.MIDDLEMAN.value, 'host'),
+                int(config.get(Components.MIDDLEMAN.value, 'port')),
+            )
+        )
+        send_over_stream(
+            connection=client_socket,
+            raw_message=golem_message_frame,
+            private_key=CONCENT_PRIVATE_KEY,
+        )
+
+        middleman_process.kill()
+
+        # Waiting for MiddleMan to finish.
+        sleep(10)
+
+        middleman_process = run_middleman()
+
+        # Waiting for MiddleMan to start.
+        sleep(10)
+
+        receive_frame_generator = unescape_stream(connection=client_socket)
+        try:
+            next(receive_frame_generator)
+        except socket.error as exception:
```

Comment:
`try/except` inside `try/except`, Bro. :P

Reply:
What is the problem with it ? ;)

Reply:
I know that @cameel is not here right now, but he's always pointing thinks like that out. `try/except` block should always catch exception that we expect to happen sometimes. I understand this part `except socket.error ...`, but I don't understand why the whole test is a `try/except` block.

Reply:
But we do it all the time :) Consider this - we use view decorators, which does `try/except`, and they call view code in `try` block, and inside view we do more `try/except`.

Actually here the `except socket.error as exception` is part of the test flow, which checks if old connection is dead after `SigningService` restart. This is the only way to verify this I think. I know it looks weird, but the whole test case is not typical one.

> but I don't understand why the whole test is a try/except block.

If anything unexpected happens during the test, `SigningService` and `MiddleMan` processed must be killed, otherwise you would have to kill them manually in console. It is done for convenience mostly.

Reply:
>  I know it looks weird, but the whole test case is not typical one.

True, okey then, I just wanted to be sure there was no other reasonable option to do this.